### PR TITLE
Winter js test suite

### PR DESCRIPTION
**Pull request by dynamite-bud**
Test Suite for running tests for front facing winterjs for APIs mentioned in #9 

Tests with:
- [x] wasmer-winter
- [x] wrangler
- [x] workerd

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -8,6 +8,9 @@ edition = "2021"
 
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
+# [workspace]
```
Comment (Arshia001) [inline Cargo.toml:11]: Remove commented code

File: test-suite/.wrangler/tmp/bundle-YQWXdC/middleware-insertion-facade.js
Comment (Arshia001) [inline test-suite/.wrangler/tmp/bundle-YQWXdC/middleware-insertion-facade.js:1]: I don't think we need the files in this folder in the repository.

File: test-suite/node_modules/.mf/cf.json
Comment (Arshia001) [inline test-suite/node_modules/.mf/cf.json:1]: Is this file necessary?

File: test-suite/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+#[derive(Debug, Clone, Deserialize)]
+pub enum ERunner {
+    Winter(Runner),
+    Wrangler(Runner),
+    WorkerD(Runner),
+}
+
+pub trait RunnerTrait {
```
Comment (Arshia001) [inline test-suite/src/lib.rs:34]: You usually name this just `Run` rather than `Runner`. See the traits already in rust, e.g. `Read` and `Write`.

File: test-suite/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+#[derive(Debug, Clone, Deserialize)]
+pub struct Runner {
+    pub name: String,
+    pub command: String,
+    pub port: u16,
+    pub args: Vec<String>,
+}
+
+#[derive(Debug, Clone, Deserialize)]
+pub enum ERunner {
```
Comment (Arshia001) [inline test-suite/src/lib.rs:28]: Now, I may be nitpicking, but I find this use of an enum very irregular. You have 3 cases, all of which store the same inner data, and the case itself isn't actually useful for anything (aside from serialization in the toml file; but that's not a good enough reason).

An enum is useful for "making illegal state unrepresentable" ([a very good explanation can be found here](https://fsharpforfunandprofit.com/posts/designing-with-types-making-illegal-states-unrepresentable/)). This enum is, indeed, creating *more* representable illegal states; for example, consider the case of `ERunner::Winter(Runner { name = "wrangler" })`.

You can easily remove this enum, which will also let you remove the trait as well. To deserialize from the toml, if you want to keep the `Winter =`, you could use a `Map<String, Runner>`. However, I also find that to be redundant; if the label is removed, the file can be reduced to:

```toml
[[matrix_runners]]
name = "winter"
...

[[matrix_runners]]
name = "wrangler"
...
```

which is more readable too.

To summarize, if you're not using an enum to store different data or as a way to create branching logic, stop and reconsider if you actually need an enum.

Comment (Arshia001) [reply test-suite/src/lib.rs:28]: I feel it'd be unfair not to mention that it would, indeed, be appropriate to use an enum here if, instead of deserializing, we were hardcoding the values, like so:

```rust
enum Runner {
  Winter,
  Wrangler,
}

impl Runner {
  fn port(&self) -> u16 {
    match self {
      Self::Winter => 8080,
      Self::Wrangler => 8787,
    }
  }
}

File: test-suite/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+    fn port(&self) -> u16 {
+        self.port
+    }
+    fn name(&self) -> String {
+        self.name.clone()
+    }
+}
+
+impl RunnerTrait for ERunner {
```
Comment (Arshia001) [inline test-suite/src/lib.rs:57]: Once the enum is removed, this block of entirely redundant code can also be removed.

File: test-suite/tests/integration.rs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+                    eprintln!("Waiting for port {}...", port);
+                    thread::sleep(Duration::from_secs(3));
+                    true
+                }
+            }
+        });
+    }
+}
+
+fn main() -> Result<(), anyhow::Error> {
```
Comment (Arshia001) [inline test-suite/tests/integration.rs:42]: nit #2: Since this entire code is meant as a test runner, we could make it a binary, move all of this to `main.rs`, and run it with `cargo run` instead of `cargo test`, which should produce similar results.

File: test-suite/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,27 @@
+[package]
```
Comment (Arshia001) [inline test-suite/Cargo.toml:1]: Is this package in a separate workspace than the top-level WinterJS code? I'm not seeing anything that would put it in a separate workspace.

File: test-suite/tests/integration.rs [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+use std::{
+    fs,
+    net::{Ipv4Addr, SocketAddr, SocketAddrV4, TcpStream},
+    thread,
+    time::Duration,
+};
+
+// Read and parse the TOML file into a TestConfig
+fn read_test_cases() -> Result<TestConfig, anyhow::Error> {
+    let file_path = format!("{}/tests/tests.toml", env!("CARGO_MANIFEST_DIR"));
```
Comment (Arshia001) [inline test-suite/tests/integration.rs:13]: You're hardcoding the path to `tests.toml` completely (as in, `/home/myself/repos/winterjs/test-suite/tests/tests.toml`), since `env!` hardcodes the value the environment variable has at *compile time* into the binary.

While it's true that this code will probably always run right after it's compiled and the executable then discarded, meaning this won't cause a problem, I still think we can do it better. Two options:

* use `include_str!` to include the test cases directly into the binary as a string. This way, it'll be simpler to put the code in other environments (such as in a WASM module).
* use a relative path; cargo always launches programs with `pwd` set to `CARGO_MANIFEST_DIR`, so you can just use `tests.toml` and put the file next to `Cargo.toml` or `tests/tests.toml` if you want to keep it here.

**Review by Arshia001 — CHANGES_REQUESTED** [review]

File: Cargo.toml (hunk 1 above)
Comment (dynamite-bud) [reply Cargo.toml:11]: Removed

File: test-suite/node_modules/.mf/cf.json
Comment (dynamite-bud) [reply test-suite/node_modules/.mf/cf.json:1]: removed

File: test-suite/.wrangler/tmp/bundle-YQWXdC/middleware-insertion-facade.js
Comment (dynamite-bud) [reply test-suite/.wrangler/tmp/bundle-YQWXdC/middleware-insertion-facade.js:1]: removed

File: test-suite/src/lib.rs (hunk 3 above)
Comment (dynamite-bud) [reply test-suite/src/lib.rs:28]: If I do this then I need to hardcode the command, port, etc.

File: test-suite/tests/integration.rs (hunk 5 above)
Comment (dynamite-bud) [reply test-suite/tests/integration.rs:42]: But won't the convention of using it as `cargo test` conform to testing ?

File: test-suite/Cargo.toml (hunk 6 above)
Comment (dynamite-bud) [reply test-suite/Cargo.toml:1]: No it's not in a separate workspace but rather it's own package

File: test-suite/tests/integration.rs (hunk 7 above)
Comment (dynamite-bud) [reply test-suite/tests/integration.rs:13]: Yeah, `include_str!` would be better

File: test-suite/src/lib.rs (hunk 3 above)
Comment (Arshia001) [reply test-suite/src/lib.rs:28]: I'm not suggesting we hardcode stuff, just saying that *if* we were going to hardcode, then an enum would have been a suitable option.

File: test-suite/Cargo.toml (hunk 6 above)
Comment (Arshia001) [reply test-suite/Cargo.toml:1]: NVM, the root `Cargo.toml` isn't a workspace.

File: test-suite/tests/integration.rs (hunk 5 above)
Comment (dynamite-bud) [reply test-suite/tests/integration.rs]: I added it to `main.rs`. Being in favour of `cargo run` with you

File: test-suite/src/lib.rs (hunk 4 above)
Comment (dynamite-bud) [reply test-suite/src/lib.rs]: Removed it

File: test-suite/src/lib.rs (hunk 3 above)
Comment (dynamite-bud) [reply test-suite/src/lib.rs]: Hardcoded it for now

**Review by Arshia001 — APPROVED** [review]
